### PR TITLE
Detect "tag" fields and disable term frequency check

### PR DESCRIPTION
**Pull request by alexksikes**
With the current default value of `min_term_freq == 2`, the common use case of
performing MLT on a tag field (a field which only contains unique values)
results in an empty result set. Instead of changing the default value of
`min_term_freq`, this commit proposes to automatically detect such a tag
field, and in this case to skip the term frequency check altogether.

### CODE REVIEW

File: src/test/java/org/elasticsearch/mlt/MoreLikeThisActionTests.java [hunk 1]
```diff
@@ -715,4 +715,30 @@ public void testMoreLikeThisIgnoreLike() throws ExecutionException, InterruptedE
...
+        logger.info("Ensuring we still return a result ...");
+        MoreLikeThisQueryBuilder mltQuery = moreLikeThisQuery("text")
+                .ids("0")
+                .include(true)
+                .minDocFreq(1)
+                .minTermFreq(randomIntBetween(2,10));
+        SearchResponse response = client().prepareSearch("test").setTypes("type1").setQuery(mltQuery).get();
+        assertSearchResponse(response);
+        assertHitCount(response, 1);
+    }
```
Comment (jpountz) [inline src/test/java/org/elasticsearch/mlt/MoreLikeThisActionTests.java]: Could this be a unit test instead of an integration test?

**Comment by jpountz** [comment]
I like this change a lot but on the other hand I don't like seeing XMoreLikeThis diverging from Lucene's MoreLikeThis. I'm good with letting this one in, but I think it should be very high priority to merge improvements to XMoreLikeThis back to Lucene and remove it from elasticsearch.

**Comment by jpountz** [comment]
> merge improvements to XMoreLikeThis back to Lucene and remove it from elasticsearch

I created #10626 to track this.

**Comment by alexksikes** [comment]
Thanks for submitting this issue. I agree!

Something to keep in mind is that the check on frequencies is only performed at the document level, not at the index. So if the field is not really a tag field, documents with all unique terms maybe yield more results than those without. The way I see it is that we try to do our best to return results in the common case of a tag field, without having to change the default on `min_term_freq`.

**Comment by alexksikes** [comment]
This is completely outdated. Will open a new PR at some point. Closing for now.

**Pull request closed without merging**